packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,307 @@
+        // The constraint can be either on the left or on the right.
+        // The constraint is on the left iff the amount required to fill the left order
+        // is less than or equal to the amount we can spend from the right order:
+        //    <leftTakerAssetAmountRemaining> <= <rightTakerAssetAmountRemaining> * <rightMakerToTakerRatio>
+        //    <leftTakerAssetAmountRemaining> <= <rightTakerAssetAmountRemaining> * <rightOrder.makerAssetAmount> / <rightOrder.takerAssetAmount>
+        //    <leftTakerAssetAmountRemaining> * <rightOrder.takerAssetAmount> <= <rightTakerAssetAmountRemaining> * <rightOrder.makerAssetAmount>
+        uint256 rightTakerAssetAmountRemaining = safeSub(rightOrder.takerAssetAmount, rightOrderFilledAmount);
+        uint256 leftTakerAssetAmountRemaining = safeSub(leftOrder.takerAssetAmount, leftOrderFilledAmount);
+        uint256 leftOrderAmountToFill = 0;
+        uint256 rightOrderAmountToFill = 0;
+        if (
+            safeMul(leftTakerAssetAmountRemaining, rightOrder.takerAssetAmount) <=
+            safeMul(rightTakerAssetAmountRemaining, rightOrder.makerAssetAmount)
+        ) {
+            // Left order is the constraint: maximally fill left
+            leftOrderAmountToFill = leftTakerAssetAmountRemaining;
+
+            // The right order receives an amount proportional to how much was spent.
+            // TODO: Can we ensure rounding error is in the correct direction?
+            rightOrderAmountToFill = safeGetPartialAmount(
+                rightOrder.takerAssetAmount,
+                rightOrder.makerAssetAmount,
+                leftOrderAmountToFill);
+        } else {
+            // Right order is the constraint: maximally fill right
+            rightOrderAmountToFill = rightTakerAssetAmountRemaining;
+
+            // The left order receives an amount proportional to how much was spent.
+            // TODO: Can we ensure rounding error is in the correct direction?
+            leftOrderAmountToFill = safeGetPartialAmount(
+                rightOrder.makerAssetAmount,
+                rightOrder.takerAssetAmount,
+                rightOrderAmountToFill);
+        }
+
+        // Calculate fill results for left order
+        (   status,
+            matchedFillResults.left
+        ) = calculateFillResults(
+            leftOrder,
+            leftOrderStatus,
+            leftOrderFilledAmount,
+            leftOrderAmountToFill);
+        if (status != uint8(Status.SUCCESS)) {
+            return (status, matchedFillResults);
+        }
+
+        // Calculate fill results for right order
+        (   status,
+            matchedFillResults.right
+        ) = calculateFillResults(
+            rightOrder,
+            rightOrderStatus,
+            rightOrderFilledAmount,
+            rightOrderAmountToFill);
+        if (status != uint8(Status.SUCCESS)) {
+            return (status, matchedFillResults);
```

Comment:
Hmm, if the left status is successful but the right is not, `matchedFillResults` will be inaccurate here.